src/new_hope/week1.clj

```diff
@@ -50,3 +50,196 @@
+
+;let
+(= 7 (let [x 5] (+ 2 x)))
+(= 7 (let [x 3, y 10] (- y x)))
+(= 7 (let [x 21] (let [y 3] (/ x y))))
+
+;let it be
+(= 10 (let [x 7, y 3] (+ x y)))
+(= 4 (let [y 3, z 1] (+ y z)))
+(= 1 (let [z 1] z))
+
+;regex
+(= "ABC" (apply str (re-seq #"[A-Z]+" "bA1B3Ce ")))
+
+;recursive function
+(= '(5 4 3 2 1) ((fn foo [x] (when (> x 0) (conj (foo (dec x)) x))) 5))
+
+;tail recursion
+(= [7 6 5 4 3]
+   (loop [x 5
+          result []]
+     (if (> x 0)
+       (recur (dec x) (conj result (+ 2 x)))
+       result)))
+
+;thread-first macro
+(= (last (sort (rest (reverse [2 5 4 1 3 6]))))
+   (-> [2 5 4 1 3 6] (reverse) (rest) (sort) (last))
+   5)
+
+;thread-last macro
+(= (apply + (map inc (take 3 (drop 2 [2 5 4 1 3 6]))))
+   (->> [2 5 4 1 3 6] (drop 2) (take 3) (map inc) (apply +))
+   11)
+
+;for
+(= '(1 5 9 13 17 21 25 29 33 37) (for [x (range 40)
+                                       :when (= 1 (rem x 4))]
+                                   x))
+(= '(1 5 9 13 17 21 25 29 33 37) (for [x (iterate #(+ 4 %) 0)
+                                       :let [z (inc x)]
+                                       :while (< z 40)]
+                                   z))
+(= '(1 5 9 13 17 21 25 29 33 37) (for [[x y] (partition 2 (range 20))]
+                                   (+ x y)))
+
+;sum it up
+(def sum-it-up
+  (fn [coll]
+    (reduce + coll)))
+
+(= (sum-it-up [1 2 3]) 6)
+(= (sum-it-up (list 0 -2 5 5)) 8)
+(= (sum-it-up #{4 2 1}) 7)
+(= (sum-it-up '(0 0 -1)) -1)
+(= (sum-it-up '(1 10 3)) 14)
+
+;find odd numbers
+(def find-odds
+  (fn [coll]
+    (filter odd? coll)))
+
+(= (find-odds #{1 2 3 4 5}) '(1 3 5))
+(= (find-odds [4 2 1 6]) '(1))
+(= (find-odds [2 2 4 6]) '())
+(= (find-odds [1 1 1 3]) '(1 1 1 3))
+
+;palindrome
+(def palindrome?
+  (fn [x]
+    (= (seq x) (reverse x))))
+
+(false? (palindrome? '(1 2 3 4 5)))
+(true? (palindrome? "racecar"))
+(true? (palindrome? [:foo :bar :foo]))
+(true? (palindrome? '(1 1 3 3 1 1)))
+(false? (palindrome? '(:a :b :c)))
+
+;duplicate each element in sequence
+(def duplicate
+  (fn [coll]
+    (reverse
+      (reduce #(cons %2 (cons %2 %1)) '() coll))))
+
+(= (duplicate [1 2 3]) '(1 1 2 2 3 3))
+(= (duplicate [:a :a :b :b]) '(:a :a :a :a :b :b :b :b))
+(= (duplicate [[1 2] [3 4]]) '([1 2] [1 2] [3 4] [3 4]))
+
+;compress a sequence (remove duplicates)
+(def compress-seq
+  (fn [coll]
+    (dedupe coll)))
+
+(= (apply str (compress-seq "Leeeeeerrroyyy")) "Leroy")
+(= (compress-seq [1 1 2 3 3 2 2 3]) '(1 2 3 2 3))
+(= (compress-seq [[1 2] [1 2] [3 4] [1 2]]) '([1 2] [3 4] [1 2]))
+
+;pack sequence
+(def pack-seq
+  (fn [coll]
+    (partition-by identity coll)))
+
+(= (pack-seq [1 1 2 1 1 1 3 3]) '((1 1) (2) (1 1 1) (3 3)))
+(= (pack-seq [:a :a :b :b :c]) '((:a :a) (:b :b) (:c)))
+(= (pack-seq [[1 2] [1 2] [3 4]]) '(([1 2] [1 2]) ([3 4])))
+
+;find every Nth item
+(def drop-nth
+  (fn [coll n]
+    (keep-indexed
+      (fn [i elem]
+        (if (not= 0 (mod (inc i) n)) elem nil)) coll)))
```

Comment:
Arguably better, you can also use `(pos? (mod (inc i) n))` instead of `(not= 0 (mod (inc i) n))`